### PR TITLE
Transaction support

### PR DESCRIPTION
**Pull request by alcaeus**
Note: contains changes from #2462 folded into a single commit; I'd prefer merging that PR before taking on this one.

This takes the work started by @klinson in #1904 and addresses some of the feedback left in that PR:

* Transactions can't be nested. This was done as MongoDB itself can't nest transactions. Running transactions in parallel instead of nested changes the outcome of nested operations that are partially rolled back, so this is a can of worms we better not open.
* Session options are no longer hard coded. Users can specify them when calling `DB::beginTransaction` or `DB::transaction` if they need to.
* The `DB::transaction` method will respect the `$attempts` argument and not invoke the callback more often, but may also abort before invoking the callback the specified number of times. This is due to the comparable logic in the MongoDB driver aborting after two minutes. I believe this is preferable to retrying for longer, especially in a web context (where two minutes is already quite long).

Note that I have not touched the docker-compose logic. Transaction tests are skipped unless run against a deployment that supports them, and I don't think it's necessary to set up a local cluster for quick tests. People who need to debug transaction tests locally should set up their own deployment and change the connection URL via the `MONGODB_URI` environment variable when running tests.

This PR was made possible by the contributions of @klinson and @levon80999 who've worked on this in the past. Commits retain co-authorship to reflect this.

### CODE REVIEW

File: src/Concerns/ManagesTransactions.php [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+     *
+     * In version 4.0, MongoDB supports multi-document transactions on replica sets.
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
+     */
+    public function beginTransaction(array $options = []): void
+    {
+        $this->getSessionOrCreate()->startTransaction($options);
+        $this->transactions = 1;
```
Comment (jmikola) [inline src/Concerns/ManagesTransactions.php:56]: Where did this property come from? I don't recall seeing it in the previous PRs and it's not declared in this trait.

As-is, this looks like you're creating a public property at runtime ([deprecated in 8.2](https://php.watch/versions/8.2/dynamic-properties-deprecated)).

File: src/Query/Builder.php [hunk 2]
```diff
@@ -720,17 +733,19 @@ public function from($collection, $as = null)
      */
     public function truncate(): bool
     {
-        $result = $this->collection->deleteMany([]);
+        $options = $this->inheritConnectionOptions();
+        $result = $this->collection->deleteMany($options);
```
Comment (jmikola) [inline src/Query/Builder.php:737]: Subtle bug that would have gone unnoticed by type checking:

```suggestion
        $result = $this->collection->deleteMany([], $options);
```

File: tests/ConnectionTest.php [hunk 3]
```diff
@@ -37,11 +37,101 @@ public function testDb()
...
+    public function testConnectionConfig(string $expectedUri, string $expectedDatabaseName, array $config): void
+    {
+        $connection = new Connection($config);
+        $client = $connection->getMongoClient();
+
+        $this->assertSame($expectedUri, (string) $client);
+        $this->assertSame($expectedDatabaseName, $connection->getMongoDB()->getDatabaseName());
+    }
+
+    public function testConnectionWithoutConfiguredDatabase(): void
```
Comment (jmikola) [inline tests/ConnectionTest.php:129]: This likely answers my previous question about the purpose of `database` in the config struct.

File: tests/TestCase.php [hunk 4]
```diff
@@ -56,8 +56,6 @@ protected function getEnvironmentSetUp($app)
         $app['config']->set('database.connections.mysql', $config['connections']['mysql']);
         $app['config']->set('database.connections.mongodb', $config['connections']['mongodb']);
         $app['config']->set('database.connections.mongodb2', $config['connections']['mongodb']);
```
Comment (jmikola) [inline tests/TestCase.php:58]: Are `database.connections.mysql` and `database.connections.mongodb2` actually used for anything?

File: tests/TransactionTest.php [hunk 5]
```diff
@@ -0,0 +1,430 @@
...
+    {
+        $this->expectException(RuntimeException::class);
+        $this->expectExceptionMessage('There is no active session.');
+
+        DB::rollback();
+    }
+
+    private function getPrimaryServerType(): int
+    {
+        return DB::getMongoClient()->getManager()->selectServer(new ReadPreference(ReadPreference::RP_PRIMARY))->getType();
```
Comment (jmikola) [inline tests/TransactionTest.php:428]: `composer.json` currently requires `mongodb/mongodb: ^1.11`, which requires `ext-mongodb: ^1.10`. If you'd like to make this more concise, you can bump the PHPLIB requirement to `^1.12` and call [`selectServer()`](https://www.php.net/manual/en/mongodb-driver-manager.selectserver.php) without any parameter for the same effect.

File: tests/config/database.php [hunk 6]
```diff
@@ -1,35 +1,22 @@
...
+                'connectTimeoutMS'         => 100,
+                'serverSelectionTimeoutMS' => 250,
+            ],
         ],
 
         'mysql' => [
             'driver' => 'mysql',
             'host' => env('MYSQL_HOST', 'mysql'),
-            'port' => $mysqlPort,
+            'port' => env('MYSQL_PORT') ? (int) env('MYSQL_PORT') : 3306,
```
Comment (jmikola) [inline tests/config/database.php:19]: This change seems unrelated to the PR.

File: tests/config/database.php [hunk 7]
```diff
@@ -1,35 +1,22 @@
...
 return [
-
     'connections' => [
-
         'mongodb' => [
             'name' => 'mongodb',
             'driver' => 'mongodb',
-            'host' => $mongoHost,
+            'dsn' => env('MONGODB_URI', 'mongodb://127.0.0.1/'),
             'database' => env('MONGO_DATABASE', 'unittest'),
```
Comment (jmikola) [inline tests/config/database.php:9]: Note my earlier suggestion to rename this to `MONGODB_DATABASE` for consistency.

File: tests/TransactionTest.php [hunk 8]
```diff
@@ -0,0 +1,430 @@
...
+        $timesRun = 0;
+
+        DB::transaction(function () use (&$timesRun): void {
+            User::where(['name' => 'klinson'])->update(['age' => 21]);
+
+            // Update user during transaction to simulate a simultaneous update
+            DB::getCollection('users')->updateOne(['name' => 'klinson'], ['$inc' => ['age' => 2]]);
+
+            $timesRun++;
+        }, 2);
```
Comment (jmikola) [inline tests/TransactionTest.php:366]: I noted you said this [fails in CI](https://github.com/jenssegers/laravel-mongodb/actions/runs/3461141290/jobs/5778375858#step:11:23), but I'm having trouble following how this is executed. I understand that `DB::transaction()` uses a facade and is going to create an active session, but which of the operations within this callback end up using the same session (i.e. updating _within_ the transaction)?

File: .github/workflows/build-ci.yml [hunk 9]
```diff
@@ -59,6 +55,13 @@ jobs:
 
         steps:
             -   uses: actions/checkout@v2
+            -   name: Create MongoDB Replica Set
+                run: |
+                    docker run --name mongodb -p 27017:27017 -e MONGO_INITDB_DATABASE=unittest --detach mongo:${{ matrix.mongodb }} mongod --replSet rs
```
Comment (jmikola) [inline .github/workflows/build-ci.yml:60]: Remember to use `--setParameter transactionLifetimeLimitSeconds=5` (or a lower value) to minimize deadlock time in CI.

File: README.md [hunk 10]
```diff
@@ -143,47 +144,36 @@ Keep in mind that these traits are not yet supported:
...
+Instead of using a connection string, you can also use the `host` and `port` configuration options to have the connection string created for you.
 
 ```php
 'mongodb' => [
     'driver' => 'mongodb',
-    'host' => ['server1', 'server2', ...],
-    ...
+    'host' => env('DB_HOST', '127.0.0.1'),
+    'port' => env('DB_PORT', 27017),
+    'database' => env('DB_DATABASE', 'homestead'),
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply README.md:167]: The `database` option is used as the default database name for models that don't specify a database name. It is taken from the configuration as above, or extracted from the connection string if the option is missing.

File: README.md [hunk 11]
```diff
@@ -979,6 +969,46 @@ If you are familiar with [Eloquent Queries](http://laravel.com/docs/queries), th
 ### Available operations
 To see the available operations, check the [Eloquent](#eloquent) section.
 
+Transactions
+-------
+Transactions require MongoDB version ^4.0 as well as deployment of replica set or sharded clusters. You can find more information [in the MongoDB docs](https://docs.mongodb.com/manual/core/transactions/)
+
+### Basic Usage
+
+Transaction supports all operations.
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply README.md:978]: Agree, removed.

File: README.md [hunk 12]
```diff
@@ -979,6 +969,46 @@ If you are familiar with [Eloquent Queries](http://laravel.com/docs/queries), th
...
+
+```php
+DB::transaction(function () {
+    User::create(['name' => 'john', 'age' => 19, 'title' => 'admin', 'email' => 'klinsonup@gmail.com']);
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply README.md:982]: That is true, updated to use `john@example.com`.

File: README.md [hunk 13]
```diff
@@ -979,6 +969,46 @@ If you are familiar with [Eloquent Queries](http://laravel.com/docs/queries), th
...
+DB::beginTransaction();
+User::create(['name' => 'john', 'age' => 19, 'title' => 'admin', 'email' => 'klinsonup@gmail.com']);
+DB::collection('users')->where('name', 'john')->update(['age' => 20]);
+DB::collection('users')->where('name', 'john')->delete();
+
+// you can commit your changes
+DB::commit();
+
+// you can also rollback them
+//DB::rollBack();
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply README.md:999]: I've removed the `rollBack` call from this example and added a separate example on how to abort a transaction.

File: README.md [hunk 14]
```diff
@@ -979,6 +969,46 @@ If you are familiar with [Eloquent Queries](http://laravel.com/docs/queries), th
...
+```
+**NOTE:** Transactions in MongoDB cannot be nested. DB::beginTransaction() function will start new transactions in a new created or existing session and will raise the RuntimeException when transactions already exist. See more in MongoDB official docs [Transactions and Sessions](https://www.mongodb.com/docs/manual/core/transactions/#transactions-and-sessions)
+```php
+// This code will raise a RuntimeException
+DB::beginTransaction();
+    User::create(['name' => 'john', 'age' => 20, 'title' => 'admin']);
+    DB::beginTransaction()
+        DB::collection('users')->where('name', 'john')->update(['age' => 20]);
+    DB::commit()
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply README.md:1008]: I fixed those and also moved the comment about a exceptions down to where the exception is raised.

File: src/Concerns/ManagesTransactions.php (hunk 1 above)
Comment (alcaeus) [reply src/Concerns/ManagesTransactions.php:56]: This is declared in the original `Connection` class: https://github.com/laravel/framework/blob/213e07ef363bb6f2355254a5a9a0da33dcd692ad/src/Illuminate/Database/Connection.php#L127. I've re-declared the property in the trait to make it more explicit, especially since the trait could be used outside of the connection class context.

File: src/Query/Builder.php (hunk 2 above)
Comment (alcaeus) [reply src/Query/Builder.php:737]: Good catch, updated.

File: tests/TransactionTest.php (hunk 8 above)
Comment (alcaeus) [reply tests/TransactionTest.php:366]: The first method call in the transactional closure updates the `user` collection within the transaction. The second call fetches the raw collection class, bypassing the internal transaction logic.

This is the same logic I'm using in the test above this one, with the small difference that in that test, the conflicting update is only done on the first execution of the callback, which means that the transaction will commit successfully after the second callback execution. In this test, I want to specifically cause the transaction to fail every time the callback is run to test whether the repetition limit is observed. This works fine locally but for some reason fails on CI.

File: tests/config/database.php (hunk 7 above)
Comment (alcaeus) [reply tests/config/database.php:9]: Good point, changed.

File: phpunit.xml.dist [hunk 15]
```diff
@@ -35,9 +38,8 @@
         </testsuite>
     </testsuites>
     <php>
-        <env name="MONGO_HOST" value="mongodb"/>
+        <env name="MONGODB_URI" value="mongodb://127.0.0.1/" />
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply phpunit.xml.dist:41]: Changed to `MONGODB_DATABASE`.

File: src/Concerns/ManagesTransactions.php [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+        return $session;
+    }
+
+    /**
+     * Use the existing or create new session and start a transaction in session.
+     *
+     * In version 4.0, MongoDB supports multi-document transactions on replica sets.
+     * In version 4.2, MongoDB introduces distributed transactions, which adds support for multi-document transactions on sharded clusters and incorporates the existing support for multi-document transactions on replica sets.
+     *
+     * @see https://docs.mongodb.com/manual/core/transactions/
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Concerns/ManagesTransactions.php:51]: Moved the `@see` to the trait itself and used your suggested wording as comment for `beginTransaction`.

File: src/Concerns/ManagesTransactions.php [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+     * Commit transaction in this session.
+     */
+    public function commit(): void
+    {
+        $this->getSessionOrThrow()->commitTransaction();
+        $this->transactions = 0;
+    }
+
+    /**
+     * Rollback transaction in this session.
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Concerns/ManagesTransactions.php:69]: I've left the comments in place for the sake of completeness, but I did change the wording to use "Abort".

File: src/Connection.php [hunk 18]
```diff
@@ -285,7 +289,7 @@ protected function getDefaultSchemaGrammar()
     /**
      * Set database.
      *
-     * @param \MongoDB\Database $db
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Connection.php:288]: Not sure why there's a mix of these, but I'd like to clean those up in a separate PR.

File: src/Connection.php [hunk 19]
```diff
@@ -295,8 +299,8 @@ public function setDatabase(\MongoDB\Database $db)
     /**
      * Dynamically pass methods to the connection.
      *
-     * @param string $method
-     * @param array $parameters
+     * @param  string  $method
+     * @param  array  $parameters
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Connection.php:303]: StyleCI suggested all these changes - not sure why they weren't applied to begin with. Unfortunately I was "smart" enough to squash the changes into my original commit, but I can undo them if you'd like.

File: src/Eloquent/Model.php [hunk 20]
```diff
@@ -402,7 +402,7 @@ public function getForeignKey()
     /**
      * Set the parent relation.
      *
-     * @param \Illuminate\Database\Eloquent\Relations\Relation $relation
+     * @param  \Illuminate\Database\Eloquent\Relations\Relation  $relation
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Eloquent/Model.php:405]: See above - I'd clean this up in a separate PR.

File: src/Query/Builder.php [hunk 21]
```diff
@@ -720,17 +733,19 @@ public function from($collection, $as = null)
...
 
         return 1 === (int) $result->isAcknowledged();
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Query/Builder.php:739]: Deferring this to a separate cleanup PR.

File: tests/TransactionTest.php (hunk 5 above)
Comment (alcaeus) [reply tests/TransactionTest.php:428]: It's a good thing that we've decided to align the version numbers of the library and the extension. The change to `selectServer()` was made in `ext-mongodb` 1.11, which is required as of `mongodb/mongodb` 1.10. Somehow the fact that we default to a primary read preference had escaped my memory. I've updated the call.

File: src/Connection.php (hunk 19 above)
[the remark this answers is not in the document]
Comment (jmikola) [reply src/Connection.php]: No worries. If that's part of the repo it's fine. I suppose folks forgot to apply it to existing source files when integrating.

File: tests/config/database.php (hunk 6 above)
Comment (jmikola) [reply tests/config/database.php]: Probably cleaner to do that, and it shouldn't conflict with anything in this PR since the lines aren't adjacent.

File: tests/TransactionTest.php (hunk 8 above)
Comment (jmikola) [reply tests/TransactionTest.php:366]: Are you running a single node replica set with the same server version (I don't think patch is relevant) locally?

Comment (alcaeus) [reply tests/TransactionTest.php]: This test was now fixed 🎉

**Comment by alcaeus** [comment]
@Smolevich thanks for your patience! This PR is now ready for review. I'll make a couple of minor changes to the CI setup to ensure that any aborted or stale transaction doesn't block the CI build for an unnecessary period of time (30 seconds by default), but the functionality is in place and ready to go!

File: .github/workflows/build-ci.yml [hunk 22]
```diff
@@ -59,6 +55,13 @@ jobs:
...
+                    until docker exec --tty mongodb mongo 127.0.0.1:27017 --eval "db.runCommand({ ping: 1 })"; do
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply .github/workflows/build-ci.yml:61]: Done - I added an additional build step for this to avoid people having to look through the entire setup output for this.

File: .github/workflows/build-ci.yml (hunk 9 above)
Comment (alcaeus) [reply .github/workflows/build-ci.yml]: Done.

File: src/Concerns/ManagesTransactions.php [hunk 23]
```diff
@@ -93,11 +95,22 @@ public function transaction(Closure $callback, $attempts = 1, array $options = [
                 return;
             }
 
-            $callbackResult = $callback($this);
+            // Catch, store and re-throw any exception thrown during execution
```
[the remark this answers is not in the document]
Comment (alcaeus) [reply src/Concerns/ManagesTransactions.php]: One day, I'll stop forgetting Oxford commas...

**Review by divine — APPROVED** [review]
Thanks!

**Comment by axgalache** [comment]
Any clue on when this feature may be released?
